prow/bugzilla/client.go

```diff
@@ -220,31 +259,105 @@ func (c *client) GetRootForClone(bug *Bug) (*Bug, error) {
 // GetAllClones returns all the clones of the bug including itself
 // Differs from GetClones as GetClones only gets the child clones which are one level lower
 func (c *client) GetAllClones(bug *Bug) ([]*Bug, error) {
-	return getAllClones(c, bug)
+	bugCache := newBugDetailsCache()
+	return getAllClones(c, bug, bugCache)
 }
-func getAllClones(c Client, bug *Bug) ([]*Bug, error) {
-	root, err := getRootForClone(c, bug)
+
+func getAllClones(c Client, bug *Bug, bugCache *bugDetailsCache) (clones []*Bug, err error) {
+
+	clones = []*Bug{}
+	err = getAllLinkedBugs(c, bug.ID, bugCache, nil)
 	if err != nil {
 		return nil, err
 	}
-	clones, err := getRecursiveClones(c, root)
-	if err != nil {
-		return nil, err
+	cachedBugs := bugCache.list()
+	for index, node := range cachedBugs {
+		if node.Summary == bug.Summary {
+			clones = append(clones, &cachedBugs[index])
+		}
 	}
-	// Getting rid of the bug whose clones we are searching for
-	// we could optimize this (maybe?) if the list turns out to be too long
-	indexOfOriginal := -1
-	for index, clone := range clones {
-		if clone.ID == bug.ID {
-			// Check if there is a more elegant way to do this
-			indexOfOriginal = index
+	return clones, nil
+}
+
+// Parallel implementation for getAllClones - spawns threads to go up and down the tree
+// Also parallelizes the getBug calls if bug has multiple bugs in DependsOn/Blocks
+func getAllLinkedBugs(c Client, bugID int, bugCache *bugDetailsCache, errGroup *errgroup.Group) error {
+	var shouldWait bool
+	if errGroup == nil {
+		shouldWait = true
+		errGroup = new(errgroup.Group)
+	}
+	var err error
+	var bug *Bug
+	bugObj, cacheHasBug := bugCache.get(bugID)
+	if !cacheHasBug {
+		bug, err = c.GetBug(bugID)
```

Comment:
just re-use the `bugObj` variable here and save yourself the allocation and the `else` branch

Reply:
bugObj is type Bug, bug is of type *Bug

Reply:
So its `*bugObj, err = c.GetBug(bugID)`

Reply:
it spits out an error message saying its an invalid indirect for the above...

Reply:
```diff
diff --git a/prow/bugzilla/client.go b/prow/bugzilla/client.go
index f5c15ee5e4..49350af88f 100644
--- a/prow/bugzilla/client.go
+++ b/prow/bugzilla/client.go
@@ -287,22 +287,19 @@ func getAllLinkedBugs(c Client, bugID int, bugCache *bugDetailsCache, errGroup *
                shouldWait = true
                errGroup = new(errgroup.Group)
        }
-       var err error
-       var bug *Bug
        bugObj, cacheHasBug := bugCache.get(bugID)
        if !cacheHasBug {
-               bug, err = c.GetBug(bugID)
+               bug, err := c.GetBug(bugID)
                if err != nil {
                        return err
                }
-       } else {
-               bug = &bugObj
+               bugObj = *bug
        }
        errGroup.Go(func() error {
-               return traverseUp(c, bug, bugCache, errGroup)
+               return traverseUp(c, &bugObj, bugCache, errGroup)
        })
        errGroup.Go(func() error {
-               return traverseDown(c, bug, bugCache, errGroup)
+               return traverseDown(c, &bugObj, bugCache, errGroup)
        })
```

Reply:
Why is it better to do it in this way?

Reply:
also bug wont be in scope for the else loop if I do that

Reply:
There is no `else` block anymore. Its better because you save the `else` block and the preacllication of `var err, bug error *Bug`